tribits/core/package_arch/TribitsReadDepsFilesCreateDepsGraph.cmake

```diff
@@ -541,7 +545,7 @@ endmacro()
 #
 # See `Function call tree for constructing package dependency graph`_.
 #
-function(tribits_set_dep_packages  packageName  testOrLib  requiredOrOptional  pkgsOrTpls)
+macro(tribits_set_dep_packages  packageName  testOrLib  requiredOrOptional  pkgsOrTpls)
```

Comment:
Be careful changing a `function` to a `macro` - this will leak every single variable used inside the macro. It might be better to `set(... PARENT_SCOPE)` on the variables you need to return instead.

Reply:
I know the issues.  The problem I have is that it is very difficult to work with global variables in any type of logical way.  When you have a set of  CMake code that is trying to coordinate to build a data-structure of top-level project regular variables (not INTERNAL cache vars), you have to jump through all types of hoops.  And if you have a stack of function calls and you want to be setting and updating a top-level project scope var, you have to propagate those manually using `set(... PARENT_SCOPE)`.  It is a terrible situation.

I guess a way to make a stack of CMake macro calls more safe would be to namespace all of the local variables with a unique prefix (e.g. based on the macro name like `tspd_` for `tribits_set_dep_packages()`) and that would avoid classes with local variables above and below the call stack.  The one exception is macro arguments where are not CMake variables (which is super confusing because `${<macroArg}` works just like for a real CMake variable).

Reply:
Is it not true all of the things we are saying about macros are also true about local variables created in `Find<Package>.cmake` files and `<Package>Config.cmake` files?  When those files get processed, are then not run in the same local variable scope as the code calling `find_package()`?   I think suggests that all of this code needs to carefully namespace local variables.

Reply:
>  though I don't see it as much effort to just add `set(... PARENT_SCOPE)` at the end.

When you have as stack of subroutines that are building a  global data-structure all of this is just so messy.  It is not hard to come up with use cases where variables don't behave the way you think they should.  I think I would just rather namespace a small number of local vars within a macro in cases like this than to fight the CMake variable scoping and hiding system (see section "5.4. Potentially Surprising Behavior Of Variables" in the book "Professional CMake").  Otherwise, I try to use functions as much as possible to avoid all of these problems (and all routines that don't modify state should be functions).